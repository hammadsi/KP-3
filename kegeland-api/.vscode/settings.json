{
  "editor.defaultFormatter": "esbenp.prettier-vscode",
  "eslint.validate": ["javascript", "typescript"],
  "editor.codeActionsOnSave": {
    "source.fixAll.eslint": true
  },
  "files.eol": "\r\n"
}
