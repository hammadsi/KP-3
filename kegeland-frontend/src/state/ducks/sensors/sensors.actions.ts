import { createAsyncThunk } from '@reduxjs/toolkit';

import { apiCaller } from '../../../utils/apiCaller';

import { Sensor, SensorType } from './sensors.interface';

export const fetchSensor = createAsyncThunk(
  'sensors/fetchSensor',
  async (sensorType: SensorType) =>
    apiCaller<Sensor>({
      url: `sensors/${sensorType}`,
      method: 'GET',
    }),
);
