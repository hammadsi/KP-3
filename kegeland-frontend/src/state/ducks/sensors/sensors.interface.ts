export enum SensorType {
  FEMFIT = 'femfit',
  EMPATICA = 'empatica',
  IMU = 'imu',
}

export type Sensor = {
  id: string;
  name: SensorType;
  labels: string[];
};

export interface SensorState {
  loading: boolean;
  data: {
    [key in SensorType]?: Sensor;
  };
  error: string | undefined;
}
