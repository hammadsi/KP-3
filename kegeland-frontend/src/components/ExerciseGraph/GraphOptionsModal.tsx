import {
  Button,
  Divider,
  Grid,
  GridItem,
  HStack,
  Modal,
  ModalBody,
  ModalContent,
  ModalFooter,
  ModalHeader,
  ModalOverlay,
  ModalProps,
  Select,
  Switch,
  Text,
} from '@chakra-ui/react';
import { ChartType } from 'chart.js';

import useGraphProfileSelector from '../../hooks/useGraphProfileSelector';
import { Sensor } from '../../state/ducks/sensors/sensors.interface';

export type GraphOptionsProps = {
  sensor: Sensor;
  updatePlot: (label: string, plot: ChartType) => void;
  resetPlot: () => void;
  updateXAxis: (value: boolean) => void;
} & Omit<ModalProps, 'children'>;

const GraphOptionsModal: React.FC<GraphOptionsProps> = ({
  sensor,
  updatePlot,
  resetPlot,
  updateXAxis,
  isOpen,
  onClose,
  ...props
}) => {
  const profile = useGraphProfileSelector(sensor.name);

  return (
    <Modal {...props} isOpen={isOpen} onClose={onClose}>
      <ModalOverlay>
        <ModalContent>
          <ModalHeader paddingTop={2.5}>
            <Text>Chart Configuration</Text>
          </ModalHeader>
          {profile ? (
            <ModalBody>
              <Text fontWeight="semibold" fontSize={18} marginBottom={2}>
                Plot settings
              </Text>
              <Grid templateColumns="repeat(2, 1fr)" gap={6}>
                {Object.entries(profile.labels).map(([label, val]) => (
                  <GridItem key={label}>
                    <HStack>
                      <Text fontWeight="medium">{label}</Text>
                      <Select
                        value={val.plotType}
                        onChange={(event) =>
                          updatePlot(label, event.target.value as ChartType)
                        }>
                        <option value="line">Line plot</option>
                        <option value="bar">Bar plot</option>
                        <option value="scatter">Scatter plot</option>
                      </Select>
                    </HStack>
                  </GridItem>
                ))}
              </Grid>
              <Divider marginY={4} />
              <Text fontWeight="semibold" fontSize={18} marginBottom={2}>
                Labels
              </Text>

              <Text fontWeight="semibold" fontSize={16}>
                X-axis
              </Text>

              <HStack>
                <Text>Display timedelta</Text>
                <Switch
                  isChecked={profile.useTimedelta}
                  onChange={() => updateXAxis(!profile.useTimedelta)}
                />
              </HStack>
            </ModalBody>
          ) : (
            <p>Loading</p>
          )}
          <ModalFooter justifyContent="space-between">
            <Button variant="ghost" onClick={resetPlot}>
              Reset to default
            </Button>
            <Button color="white" onClick={onClose}>
              Close
            </Button>
          </ModalFooter>
        </ModalContent>
      </ModalOverlay>
    </Modal>
  );
};

export default GraphOptionsModal;
