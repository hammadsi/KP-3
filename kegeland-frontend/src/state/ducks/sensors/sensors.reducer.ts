import { createSlice } from '@reduxjs/toolkit';

import {
  isPendingAction,
  isFulfilledAction,
  isRejectedAction,
} from '../../../utils/thunk.utils';

import { fetchSensor } from './sensors.actions';
import { SensorState } from './sensors.interface';

export const initialState: SensorState = {
  loading: false,
  data: {},
  error: undefined,
};

const sensorSlice = createSlice({
  name: 'sensors',
  initialState,
  reducers: {
    clearSensorsState: (state) => {
      state.loading = false;
      state.data = {};
      state.error = undefined;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchSensor.fulfilled, (state, { payload }) => {
        state.data[payload.name] = payload;
      })
      .addMatcher(
        (action) => isPendingAction(action, 'sensors'),
        (state) => {
          state.loading = true;
          state.error = undefined;
        },
      )
      .addMatcher(
        (action) => isFulfilledAction(action, 'sensors'),
        (state) => {
          state.loading = false;
          state.error = undefined;
        },
      )
      .addMatcher(
        (action) => isRejectedAction(action, 'sensors'),
        (state, { error }) => {
          state.loading = false;
          state.error = error.message;
        },
      );
  },
});

export const { clearSensorsState } = sensorSlice.actions;

export const sensorReducer = sensorSlice.reducer;
