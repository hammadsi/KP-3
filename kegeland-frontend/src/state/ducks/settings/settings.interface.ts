import { SensorType } from '../sensors/sensors.interface';

export type GraphLabels = {
  [key: string]: { plotType: 'bar' | 'line' | 'scatter'; hidden: boolean };
};

export type GraphProfile = {
  useTimedelta: boolean;
  labels: GraphLabels;
};

export type SettingsState = {
  graph: {
    [key in SensorType]: GraphProfile | undefined;
  };
};
