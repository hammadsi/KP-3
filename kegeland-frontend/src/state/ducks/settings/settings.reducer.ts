import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { SensorType } from '../sensors/sensors.interface';

import { GraphProfile, SettingsState } from './settings.interface';

export const initialState: SettingsState = {
  graph: {
    femfit: undefined,
    empatica: undefined,
    imu: undefined,
  },
};

const settingsSlice = createSlice({
  name: 'settings',
  initialState,
  reducers: {
    setGraphProfile: (
      state: SettingsState,
      action: PayloadAction<{
        sensor: SensorType;
        profile: GraphProfile | undefined;
      }>,
    ) => {
      const { sensor, profile } = action.payload;
      state.graph[sensor] = profile;
    },
  },
});

export const { setGraphProfile } = settingsSlice.actions;

export default settingsSlice.reducer;
