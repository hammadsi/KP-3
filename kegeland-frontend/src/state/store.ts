import { combineReducers, configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import authReducer from './ducks/auth/auth.reducer';
import patientsReducer from './ducks/patients/patients.reducer';
import questionnairesReducer from './ducks/questionnaires/questionnaires.reducer';
import { sensorReducer } from './ducks/sensors/sensors.reducer';
import sessionsReducer from './ducks/sessions/sessions.reducer';
import settingsReducer from './ducks/settings/settings.reducer';
import wheelchairPatientsReducer from './ducks/wheelchairPatients/wheelchairPatients.reducer';

const persistConfig = {
  key: 'root',
  version: 1,
  storage,
  whitelist: ['settings', 'auth'],
};

export const rootReducer = combineReducers({
  settings: settingsReducer,
  auth: authReducer,
  patients: patientsReducer,
  sensors: sensorReducer,
  sessions: sessionsReducer,
  questionnaires: questionnairesReducer,
  wheelchairPatients: wheelchairPatientsReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const isDev = !['production', 'test'].includes(process.env.NODE_ENV || '');

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    isDev
      ? getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }).concat(logger)
      : getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }),
  devTools: isDev,
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
