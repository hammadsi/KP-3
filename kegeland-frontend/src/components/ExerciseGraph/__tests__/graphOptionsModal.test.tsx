import { merge } from 'lodash';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

import { SensorType } from '../../../state/ducks/sensors/sensors.interface';
import graphProfileMocks from '../../../state/ducks/settings/mocks/graphProfile.mocks';
import { SettingsState } from '../../../state/ducks/settings/settings.interface';
import { initialState } from '../../../state/ducks/settings/settings.reducer';
import { initialStore, mockStore } from '../../../state/mocks/store.mock';
import GraphOptionsModal from '../GraphOptionsModal';

describe('Test graph options modal', () => {
  const settingsState: SettingsState = {
    ...initialState,
    graph: {
      [SensorType.FEMFIT]: graphProfileMocks,
      [SensorType.EMPATICA]: undefined,
      [SensorType.IMU]: undefined,
    },
  };

  const store = mockStore(merge({ settings: settingsState }, initialStore));
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <GraphOptionsModal
            sensor={{
              id: 'sensorId',
              name: SensorType.FEMFIT,
              labels: ['s1', 's2', 's3'],
            }}
            updatePlot={jest.fn()}
            resetPlot={jest.fn()}
            updateXAxis={jest.fn()}
            isOpen={false}
            onClose={jest.fn()}
          />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
