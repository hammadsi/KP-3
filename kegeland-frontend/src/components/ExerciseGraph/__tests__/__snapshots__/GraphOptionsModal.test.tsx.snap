// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test graph options modal renders correctly 1`] = `null`;
